prow/tide/tide.go

```diff
@@ -866,46 +865,108 @@ func (c *changedFilesAgent) prune() {
+	toTrigger map[string]config.Presubmit
+}
+
+func (c *Controller) requirementsByPull(sp *subpool) (map[int]prJobRequirements, error) {
+	reqs := make(map[int]prJobRequirements, len(sp.prs))
+	err := c.buildPRRequirements(reqs, sp.branch, nil, sp.prs, nil, c.ca.Config().Presubmits[sp.org+"/"+sp.repo])
+	return reqs, err
+}
+
+func (c *Controller) buildPRRequirements(m map[int]prJobRequirements, branch string, contextStack sets.String, prs []PullRequest, rootJob *config.Presubmit, jobs []config.Presubmit) error {
+	record := func(num int, reqJob config.Presubmit) {
+		// Allow this job context and all parent job contexts to be pending.
+		// Ensure that the root job is scheduled to trigger.
+
+		reqs, ok := m[num]
+		if !ok {
+			reqs = prJobRequirements{
+				requiredPJContexts:     sets.NewString(),
+				allowedPendingContexts: sets.NewString(),
+				toTrigger:              make(map[string]config.Presubmit),
+			}
 		}
+		reqs.requiredPJContexts.Insert(reqJob.Context)
+		reqs.allowedPendingContexts.Insert(reqJob.Context)
+		reqs.allowedPendingContexts.Insert(contextStack.List()...)
+		toTrigger := rootJob
+		if toTrigger == nil {
+			toTrigger = &reqJob
+		}
+		reqs.toTrigger[toTrigger.Name] = *toTrigger
+
+		m[num] = reqs
 	}
 
-	for _, ps := range c.ca.Config().Presubmits[sp.org+"/"+sp.repo] {
-		if !ps.ContextRequired() || !ps.RunsAgainstBranch(sp.branch) {
+	for _, ps := range jobs {
+		if !ps.RunsAgainstBranch(branch) {
 			continue
 		}
+		// Does this job require a context or have kids that might need one?
+		if rootJob == nil && !ps.AlwaysRun && ps.RunIfChanged == "" {
+			continue // No
+		}
 
-		if ps.AlwaysRun {
-			// Every PR requires this job.
-			for _, pr := range sp.prs {
-				record(int(pr.Number), ps.Context)
-			}
-		} else if ps.RunIfChanged != "" {
-			// This is a run if changed job so we need to check if each PR requires it.
-			for _, pr := range sp.prs {
+		var matchingPRs []PullRequest
+		if ps.RunIfChanged != "" {
+			// This is a run if changed job so we need to check if each PR runs it.
+			for _, pr := range prs {
 				changedFiles, err := c.changedFiles.prChanges(&pr)
 				if err != nil {
-					return nil, err
+					return err
 				}
 				if ps.RunsAgainstChanges(changedFiles) {
-					record(int(pr.Number), ps.Context)
+					// This PR runs this job.
+					matchingPRs = append(matchingPRs, pr)
+
+					if ps.ContextRequired() {
+						record(int(pr.Number), ps)
+					}
+				}
+			}
+		} else {
+			// Every PR runs this job.
+			matchingPRs = prs
+			if ps.ContextRequired() {
+				for _, pr := range prs {
+					record(int(pr.Number), ps)
 				}
 			}
 		}
+
+		if len(matchingPRs) > 0 && len(ps.RunAfterSuccess) > 0 {
+			rj := rootJob
+			if rj == nil {
+				rj = &ps
+			}
+			contexts := contextStack.Union(sets.NewString(ps.Context))
+			if err := c.buildPRRequirements(m, branch, contexts, matchingPRs, rj, ps.RunAfterSuccess); err != nil {
```

Comment:
`buildPRRequirements()` has a weird duality in API -- with and without the context stack/root job. Feels like we could make that a little clearer, it's not immediately obvious for instance in `record()` why it's valid to have a `nil` root job

Reply:
Is adding some comments to clarify sufficient? I'm not quite sure what you are requesting.

Reply:
We either pass in valid objects and use them or pass `nil` and check for it, which smells like two different functions that are currently hidden underneath one dispatch

Reply:
The context stack isn't handled any differently or checked for nil. We use a nil/empty value because the stack needs to start empty at the start of the first recursive call, not as a special case.
The root job does have a special case though (if it is nil we default it to the job we are recording). I'll change things around so that we provide the value when calling `buildPRRequirements` by adding an extra loop to `requirementsByPull`.